code/obj/machinery/spaceheater.dm

```diff
@@ -8,216 +8,180 @@ TYPEINFO(/obj/machinery/space_heater)
-					var/max = src.emagged ? 400 : 90
-					var/min = src.emagged ? -120 : 90
-
-					set_temperature = clamp(value, -min, max)
-
-				if("temp")
-					var/value = text2num_safe(href_list["val"])
-					var/max = src.emagged ? 400 : 90
-					var/min = src.emagged ? -120 : 90
-
-					// limit to 20-90 degC
-					set_temperature = clamp(set_temperature + value, -min, max)
-
-				if("cellremove")
-					if(open && cell && !usr.equipped())
-						cell.UpdateIcon()
-						usr.put_in_hand_or_drop(cell)
-						cell = null
-
-						usr.visible_message(SPAN_NOTICE("[usr] removes the power cell from \the [src]."), SPAN_NOTICE("You remove the power cell from \the [src]."))
-
-
-				if("cellinstall")
-					if(open && !cell)
-						var/obj/item/cell/C = usr.equipped()
-						if(istype(C))
-							usr.drop_item()
-							cell = C
-							C.set_loc(src)
-							C.add_fingerprint(usr)
-
-							usr.visible_message(SPAN_NOTICE("[usr] inserts a power cell into \the [src]."), SPAN_NOTICE("You insert the power cell into \the [src]."))
-
-			updateDialog()
-		else
-			usr.Browse(null, "window=spaceheater")
-			src.remove_dialog(usr)
-		return
-
+		switch(action)
+			if ("switch_off")
+				if (src.cell == null)
+					return
+				else if (src.on && src.emagged)
+					usr.show_text("The button seems to be stuck!", "red")
+				else
+					src.on = FALSE
+				playsound(src.loc, 'sound/machines/heater_off.ogg', 50, 1)
+				UpdateIcon()
+				. = TRUE
 
+			if("switch_on")
+				if (src.cell == null)
+					return
+				else if (!src.on)
+					src.on = TRUE
+					playsound(src.loc, 'sound/machines/heater_on.ogg', 50, 1)
+				UpdateIcon()
+				. = TRUE
+
+			if("set_temp")
+				var/adjust_temperature = params["temperature_adjust"]
+				var/inputted_temperature = params["inputted_temperature"] // For dragging
+				var/max = src.emagged ? 1000 : 328
+				var/min = src.emagged ? 0 : 258
+				if (adjust_temperature)
+					src.set_temperature = clamp((src.set_temperature + adjust_temperature), min , max)
+				else if (text2num_safe(inputted_temperature) != null)
+					src.set_temperature = clamp(text2num_safe(inputted_temperature), min , max)
+				. = TRUE
+
+			if("cellremove")
+				var/obj/item/I = src.cell
+				if (!I)
+					boutput(usr, SPAN_ALERT("No cell found to eject."))
+					return
+				else if (src.set_temperature > 400 && src.on)
+					usr.TakeDamage("All", 0, 15, 0, DAMAGE_BURN)
+					usr.show_text("The [src.cell.name] scalds your hand!", "red")
+					usr.emote("scream")
+					playsound(src.loc, 'sound/machines/engine_grump4.ogg', 50, 1)
+				else if (src.set_temperature < 100 && src.on)
+					usr.show_text("The [src.cell.name] seems frozen in place!", "red")
+					return
+				I.set_loc(src.loc)
+				src.on = FALSE
+				usr.put_in_hand_or_eject(I)
+				if (src.emagged)
+					usr.visible_message(SPAN_ALERT("[usr] removes the power cell from \the [src] and completely frying it!"),
+					SPAN_ALERT("You remove the power cell from \the [src], causing it to fry itself!"))
```

Comment:
Flavor text wise: I felt like the player needed some more context to what was actually going on, since the HVAC frying itself is sudden and unexpected, and giving only a visual indication that 'something went wrong' felt like too little context.
The skull icon: Yeah I can definitely work on that, feels a bit rushed. Probably removing the flavor text and replacing the skull icon with a warning of some kind of error that has occured would be a much better solution. Will work on that tommorow since its a bit late.

Reply:
Right but, why does it fry itself? Afaik it didn't used to do this.

Reply:
Your right, it didnt do this before. I implemented it only because it was wierd for the space HVAC to be completely fine after glitching out and outputting temperatures way beyond its limit, temperatures that would probably melt/overwork several vital components to never be usable again, since when emagged its heating/cooling power is 10 times more powerfull. It just didnt feel right to have the HVAC seem to be overworking only for it to be completely fine. I would have only added the death to when its overheating, but I didnt wanna give varied end-results which could lead to confusion as to why overheating gets special treatment when there are arguments for supercooling or generic operation breaking down as well.